### DISCUSSION TITLE
How to access H value in HSV block output?

### DISCUSSION THREAD

**Discussion by kshishar** [pybricks/support#1660] [Q&A]
I am using Pybricks blocks and trying use HSV mode of the color sensor to detect a color. While I am able to print the value by placing this block in the print block, I am unable to access the H value in the color object returned by the block. Is there a block or some operation I need to perform to be able to access it in block based coding?

**Comment by BertLindeman** [comment]
Use the hsv values as a list like this:

![pybricks_program_2024-06-03T06_39_07](https://github.com/pybricks/support/assets/8142081/9e01dbb6-b934-4d3c-a840-5fad507676cd)

At the moment no color sensor at hand, so I did not test.

Bert

**Reply by laurensvalk** [reply]
This doesn't work, but I think we should make it work as well as the following `h, s, v = sensor.hsv`.

This has been requested a few times, so it is probably worth doing.

**Comment by laurensvalk** [comment]
To detect colors with your own HSV values, the recommended approach is illustrated in [this video](https://www.youtube.com/watch?v=kwnSbnlhOkA). This makes color detection very reliable.

If you still want to the individual `h` value also, let me know and I'll put an example together :)

**Reply by kshishar** [reply]
Thanks! The video you linked above was helpful. However, I am trying to detect an object on the side (as opposed right under the robot which would limit the distance between sensor and the object) and when the distance between the colored object and the sensor goes more than 1-1.5 inch, the approach from the video didn't work in detecting the object. However, the H value was still within a certain range of that color, that if we used the Hue range to detect this color then it would work. What approach would you suggest here?

Also, I was looking for a video that would describe how we may mix blocks and python code. Does such a video (or documentation) exist?

**Reply by laurensvalk** [reply]
> However, I am trying to detect an object on the side (as opposed right under the robot which would limit the distance between sensor and the object) and when the distance between the colored object and the sensor goes more than 1-1.5 inch, the approach from the video didn't work in detecting the object

Just curious - did you also include the "background" color in the set? Instead of train track colors, here it would be the void, so probably just the existing `Color.NONE` included in the matching colors, unmodified.

If manual detection with the H value is going to work, I would hope the builtin detection works also, since it does something similar (if not, I'd love to improve it).

> Also, I was looking for a video that would describe how we may mix blocks and python code. Does such a video (or documentation) exist?

We have a few articles (like [this one](https://pybricks.com/project/control-the-duplo-train/)) that use this technique to enable extra features, but not yet a dedicated article on the technique itself. This is a good reminder to make one :smile: 

In your case, you could write a function like this:

```python
def get_hue(color):
    return color.h
```
Then, import the `get_hue` into your block program just like the duplo train example functions referenced above. Then you can use your new block to get the h value from a color object like the `hsv` block, or any other color object.

**Reply by kshishar** [reply]
Thanks again for the quick reply! I tried the approach and got error. Attached are the snapshots. Am I missing something or do we need a different approach here?
![image](https://github.com/pybricks/support/assets/46855734/1a53fe10-58c2-4c07-992d-a04c6dace930)
![image](https://github.com/pybricks/support/assets/46855734/d5e9de23-249d-4f7c-bdcc-590b7bb976f6)
![image](https://github.com/pybricks/support/assets/46855734/2008b0dd-b2ce-461c-b9a6-b3c361316e2f)

**Reply by laurensvalk** [reply]
If I had to guess, that's not the entire block code, but you're using multitasking elsewhere? Either using an extra start block, or with the multitask block?

In that case, the editor will assume that your Python functions are "multitask-compatible" also. To do that, you can change your module code to:

```python
async def get_hue(color):
    return color.h
```

It's also better to also remove everything unused from that file, especially the `hub` part, so you don't accidentally set up the hub twice. (That's not bad, but it can be confusing if you try to use the hub in your block code also.)

Here is minimal working example:

![image](https://github.com/pybricks/support/assets/12326241/be7747e7-8d38-4006-ac1b-e60ec3de5045)

-----------------------

I can see this isn't incredibly intuitive (to say the least :smile:), so I can see a few potential improvements from our side:

- Make `hue` easily accessible. I've opened https://github.com/pybricks/support/issues/1661 as a task.
- Check why the usual color calibration method wouldn't be sufficient (I'm still curious about that one).
- Allow you to specify whether an imported Python function is "async" or not, so you don't have to change your Python code if you change your block code.
